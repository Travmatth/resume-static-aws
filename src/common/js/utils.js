/* @flow */
import { WEEK, MONTH } from './constants';
import ExtendableError from 'extendable-error-class';

const removeChildren = (el: HTMLElement) => {
  while (el.lastChild) {
    el.removeChild(el.lastChild);
  }
};

//insert npm joke here
const padLeft = (number: string, columns: number) => {
  while (number.length < columns) {
    number = `0${number}`;
  }
  return number;
};

const parseTimeToText = (elapsedTime: number) => {
  const time = new Date(elapsedTime);
  const minutes = padLeft(time.getMinutes().toString(), 2);
  const seconds = padLeft(time.getSeconds().toString(), 2);
  const milliseconds = padLeft(time.getMilliseconds().toString(), 3);
  return `${minutes}:${seconds}.${milliseconds}`;
};

const SEC_IN_MINUTES = 60;
const kSEC_IN_SECONDS = 1000;

const scaleIntToMinutes = (val: number) =>
  val * kSEC_IN_SECONDS * SEC_IN_MINUTES;

const shrinkMinutesToInt = (val: number) =>
  val / kSEC_IN_SECONDS / SEC_IN_MINUTES;

const serialize = (url: string, params: ?Object) => {
  const query = [];

  if (params !== undefined) {
    url += '?';

    for (var property in params) {
      if (params.hasOwnProperty(property)) {
        query.push(
          encodeURIComponent(property) +
            '=' +
            encodeURIComponent(params[property]),
        );
      }
    }
  }

  return url + query.join('&');
};

class ResponseError extends ExtendableError {
  response: Response;

  constructor(message: string, response: Response) {
    super(message);
    this.response = response;
  }
}

const appendSuffix = (day: number) => {
  return day === 1
    ? `${day}st`
    : day === 2 ? `${day}nd` : day === 3 ? `${day}rd` : `${day}th`;
};

const dateString = (time: Date) => {
  return (
    `${WEEK[time.getDay()]}, ` +
    `${MONTH[time.getMonth()]} ` +
    `${appendSuffix(time.getDate())}`
  );
};

const checkForNegativeZero = (n: number) =>
  (n === 0 ? ((n = +n) || 1 / n) < 0 ? 0 : n : n);

const convertFahrenheitToCelsius = (temp: number) =>
  checkForNegativeZero(Math.round((temp - 32) * 5 / 9));

const checkHeaders = (response: Response) => {
  document.res = response;
  if (response.status !== 200)
    throw new ResponseError('fetch failed', response);
  return ((response.json(): any): Promise<any>);
};

const eventType = () => ('ontouchstart' in window ? 'touchstart' : 'click');

const trim = (str: string) => str.replace(/^\s+|\s+$/g, '');

const rand = (range: numer) => Math.floor(Math.random() * (range + 1));

const timedFetch = (promise: Promise<any>, ms: number) =>
  new Promise((resolve, reject) => {
    const id = setTimeout(() => {
      clearTimeout(id);
      reject(new Error('timeout'));
    }, ms);

    promise.then(
      response => {
        clearTimeout(id);
        resolve(response);
      },
      error => {
        clearTimeout(id);
        reject(error);
      },
    );
  });

export {
  checkForNegativeZero,
  eventType,
  checkHeaders,
  padLeft,
  parseTimeToText,
  scaleIntToMinutes,
  shrinkMinutesToInt,
  serialize,
  ResponseError,
  appendSuffix,
  dateString,
  convertFahrenheitToCelsius,
  trim,
  removeChildren,
  rand,
  timedFetch,
};
