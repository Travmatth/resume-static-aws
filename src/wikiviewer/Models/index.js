/* @flow */

import { ENDPOINT, PARAMS } from './Constants';

export { ENDPOINT, PARAMS };
