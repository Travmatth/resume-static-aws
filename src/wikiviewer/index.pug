extends ../common/pug/layout

block heading
	if (htmlWebpackPlugin)
		link(href=`${htmlWebpackPlugin.files.css}` rel="stylesheet" type="text/css")

block content
	section.hero
		include ../common/pug/nav.pug
		div.hero-body.shift-down
			div.columns
				div.column.is-10.is-offset-1
					div#center-panel.columns
						div.column
							input#search-text.has-text-centered
						div.column.is-narrow
							div#center-panel.level
								div.level-item
									button#search-btn Search
									button#random-btn Random
			div.columns.error.hidden
				div.column
					div.level
						div.level-item
							div An Error has occurred. Please check your internet connection and try again
			div.columns.spinner.hidden
				img.column.is-6.is-offset-3(src=require('assets/icons/spinner.svg'), style="margin: 10vh auto 0 auto;")
			div#content.columns.is-mobile.is-multiline.hidden
