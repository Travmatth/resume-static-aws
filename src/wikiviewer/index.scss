@import "../common/scss/_all.scss";
@import "../common/scss/_nav.scss";

.hidden {
	display: none !important;
	visibility: hidden !important;
}

.bottom-spacing {
	margin-bottom: 17px !important;
}

button:hover {
	font-size: $size-7;
}

.pull-right-desktop {}

.card {
	margin: 2px 0 2px 0;

	&:hover {
		border-left: 3px solid $primary;
	}
}

.shift-down {
	margin-top: 15vh;
}

input {
	border: solid 1px black;
}

a {
	text-decoration: underline;
	&:hover {
		color: $primary;
	}
}

.baseline {
	align-items: baseline;
}

.center-panel {
	background-color: blue;
}

input {
	width: 100%;
}

.hero-head * {
	color: black !important;
}

.full * {
	background: transparent !important;
}
