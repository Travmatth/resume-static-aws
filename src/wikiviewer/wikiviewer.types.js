/* @flow */

export type Paragraphs = HTMLCollection<HTMLParagraphElement>;
export type Headings = HTMLCollection<HTMLElement>;

export type WikiPage = {
  'pageid': number,
  'ns': number,
  'title': string,
  'index': number,
  'extract': string,
};

export type Searches = ?Array<WikiPage>;

export type WikiContinue = {
  'gsroffset': 10,
  'continue': 'gsroffset||',
};

export type Limits = {
  extracts: number,
};

export type WikiPages = {
  [string]: WikiPage,
};

export type QueryResult = {
  'pages': WikiPages,
};

export type WikiSearchResult = {|
  'batchcomplete': '' | true,
  'continue': WikiContinue,
  'query': QueryResult,
  'limits': Limits,
|};
