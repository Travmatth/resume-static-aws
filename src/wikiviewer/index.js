/* @flow */

import {
  keypressHandler,
  searchHandler,
  randomHandler,
  showScene,
} from './Handlers';
import { eventType } from 'common/js/Utils';

if (typeof document !== 'undefined')
  document.addEventListener('DOMContentLoaded', () => {
    // Anchor point for appending the result tiles
    const node = document.querySelector('#content');
    const type = eventType();

    const spinner = document.querySelector('.spinner');
    const content = document.querySelector('#content');
    const error = document.querySelector('.error');
    const show = showScene(content, spinner, error);

    document.querySelector('input').onkeypress = keypressHandler(node, show);

    document
      .getElementById('random-btn')
      .addEventListener(type, randomHandler(window));

    document
      .getElementById('search-btn')
      .addEventListener(type, searchHandler(node, show));
  });
