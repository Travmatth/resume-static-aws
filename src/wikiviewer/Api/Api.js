/* @flow */

import { ENDPOINT, PARAMS } from '../Models';
import type { WikiSearchResult, WikiPage } from '../wikiviewer.types';
import { WIKI_PROXY } from 'protected/proxies';
import { serialize, checkHeaders, withTimeout } from 'common/js/utils';

const WIKI_TIMEOUT = 5000;

const headers = new Headers({ 'Content-Type': 'text/plain' });

const processWikis = ({ query: { pages } }: WikiSearchResult) => {
  const { limits, ...wikis } = pages;
  return Object.values(wikis);
};

const search = (query: ?string): Promise<Array<WikiPage>> => {
  PARAMS['gsrsearch'] = query;

  const opts = {
    headers,
    method: 'POST',
    body: serialize(ENDPOINT, PARAMS),
  };

  return withTimeout(fetch(WIKI_PROXY, opts), WIKI_TIMEOUT)
    .then(checkHeaders)
    .then(processWikis)
    .catch(message => {
      console.error(message);
      return null;
    });
};

export { search, processWikis, WIKI_TIMEOUT };
