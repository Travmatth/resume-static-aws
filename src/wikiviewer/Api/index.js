/* @flow */

import { search, checkHeaders, processWikis } from './Api';

export { search, checkHeaders, processWikis };
