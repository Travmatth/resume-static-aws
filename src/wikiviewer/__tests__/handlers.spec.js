/* @flow */
/* eslint-env jest */

import {
  refreshResults,
  randomHandler,
  searchHandler,
  keypressHandler,
  updateDOM,
  showScene,
} from '../Handlers';
import * as Api from '../Api';
import type { Headings, Paragraphs } from '../wikiviewer.types';
import { json } from 'tests/utils';
import { wikis } from './wikiviewer.mockdata';

jest.mock('../Api', () => ({
  search: () => require('./wikiviewer.mockdata').wikis,
}));

describe('WikiViewer Handlers', () => {
  beforeEach(
    () =>
      ((document.body: any): HTMLElement).innerHTML = require('../index.pug'),
  );

  afterEach(() => {
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  it('showScene should handle default state', () => {
    const content = document.createElement('div');
    const spinner = document.createElement('div');
    const error = document.createElement('div');

    showScene(content, spinner, error)();

    expect(spinner.classList.contains('hidden')).toBe(true);
    expect(content.classList.contains('hidden')).toBe(true);
    expect(error.classList.contains('hidden')).toBe(true);
  });

  it('showScene should handle loading state', () => {
    const content = document.createElement('div');
    const spinner = document.createElement('div');
    const error = document.createElement('div');

    showScene(content, spinner, error)('loading');

    expect(content.classList.contains('hidden')).toBe(true);
    expect(spinner.classList.contains('hidden')).toBe(false);
    expect(error.classList.contains('hidden')).toBe(true);
  });

  it('showScene should handle content state', () => {
    const content = document.createElement('div');
    const spinner = document.createElement('div');
    const error = document.createElement('div');

    showScene(content, spinner, error)('content');

    expect(error.classList.contains('hidden')).toBe(true);
    expect(spinner.classList.contains('hidden')).toBe(true);
    expect(content.classList.contains('hidden')).toBe(false);
  });

  it('showScene should handle error state', () => {
    const content = document.createElement('div');
    const spinner = document.createElement('div');
    const error = document.createElement('div');

    showScene(content, spinner, error)('error');

    expect(error.classList.contains('hidden')).toBe(false);
    expect(spinner.classList.contains('hidden')).toBe(true);
    expect(content.classList.contains('hidden')).toBe(true);
  });

  it('randomHandler should change window location', () => {
    const win = { location: '' };
    randomHandler(win)();
    expect(win.location).toBe('https://en.wikipedia.org/wiki/Special:Random');
  });

  it('searchHandler should call refreshResults and updateDOM', async () => {
    const node = document.createElement('div');
    const show = jest.fn();
    document.querySelector('input').value = 'a';

    await searchHandler(node, show)();

    const a = node.querySelector('a');
    const p = node.querySelector('p');
    expect(show).toHaveBeenCalledTimes(2);
    expect(a.textContent).toBe('Big Boss');
    expect(a.href).toBe('https://en.wikipedia.org/?curid=11937256');
    expect(p.textContent).toBe('Big Boss disambiguation page.');
  });

  it('refreshResults should return early on empty query', async () => {
    const query = '';
    const node = document.createElement('div');
    const show = jest.fn();

    refreshResults(query, node, show);
    expect(show).not.toHaveBeenCalled();
  });

  it('refreshResults should call updateDOM with results of search', async () => {
    const node = document.createElement('div');
    const show = jest.fn();
    document.querySelector('input').value = 'a';

    await searchHandler(node, show)();

    const a = node.querySelector('a');
    const p = node.querySelector('p');
    expect(show).toHaveBeenCalledTimes(2);
    expect(a.textContent).toBe('Big Boss');
    expect(a.href).toBe('https://en.wikipedia.org/?curid=11937256');
    expect(p.textContent).toBe('Big Boss disambiguation page.');
  });

  it("keyPressHandler should call refreshResults on 'Enter' key", async () => {
    const query = ['foo'];
    const paragraphs = ((document.querySelectorAll('p'): any): Paragraphs);
    const headings = ((document.querySelectorAll(
      'div.heading',
    ): any): Headings);

    const handler = keypressHandler(query, headings, paragraphs);
    await handler((({ key: 'Enter' }: any): Event));

    expect(((document.body: any): HTMLElement).outerHTML).toMatchSnapshot();
  });

  it('updateDOM should create elements with returned data', async () => {
    const node = document.createElement('div');
    const show = jest.fn();
    updateDOM(wikis.slice(0, 1), node, show);
    const a = node.querySelector('a');
    const p = node.querySelector('p');

    expect(show).toHaveBeenCalledTimes(1);
    expect(a.textContent).toBe('Big Boss');
    expect(a.href).toBe('https://en.wikipedia.org/?curid=11937256');
    expect(p.textContent).toBe('Big Boss disambiguation page.');
  });

  it('updateDOM should remove previously attached elements', async () => {
    const show = jest.fn();
    const node = document.createElement('div');
    const node2 = document.createElement('div');
    node2.textContent = 'test';
    node.appendChild(node2);

    updateDOM(wikis.slice(0, 1), node, show);

    expect(show).toHaveBeenCalledTimes(1);
    expect(node.children[0].textContent).not.toBe('test');
  });
});
