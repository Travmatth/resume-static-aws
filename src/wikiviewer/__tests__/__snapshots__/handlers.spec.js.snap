// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WikiViewer Handlers keyPressHandler should call refreshResults on 'Enter' key 1`] = `"<body><meta charset=\\"utf-8\\"><meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\"><meta name=\\"description\\" content=\\"\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, height=device-height\\"><div class=\\"shell\\"><section class=\\"hero\\"><div class=\\"hero-head fixed\\"><header class=\\"navbar\\"><div class=\\"navbar-brand\\"><a class=\\"navbar-item\\" href=\\"/\\">Home</a><a class=\\"navbar-item\\" id=\\"projects-btn\\">Projects</a><a class=\\"navbar-item\\" href=\\"https://medium.com/@travmath\\" target=\\"_blank\\">Blog</a></div></header><div><header class=\\"nav\\"><div class=\\"container\\"><nav class=\\"tabs is-boxed full\\"><div class=\\"container dropdown-custom full is-fullwidth container\\"><ul><li class=\\"blur is-marginless\\"><a href=\\"/calculator\\">calculator</a></li><li class=\\"blur is-marginless\\"><a href=\\"/localweather\\">weather</a></li><li class=\\"blur is-marginless\\"><a href=\\"/pomodoro\\">pomodoro</a></li><li class=\\"blur is-marginless\\"><a href=\\"/randomquote\\">quote</a></li><li class=\\"blur is-marginless\\"><a href=\\"/simon\\">simon</a></li><li class=\\"blur is-marginless\\"><a href=\\"/tictactoe\\">tic-tac-toe</a></li><li class=\\"blur is-marginless\\"><a href=\\"/twitchtv\\">twitch</a></li><li class=\\"blur is-marginless\\"><a href=\\"/wikiviewer\\">wiki</a></li></ul></div></nav></div></header></div></div><div class=\\"hero-body shift-down\\"><div class=\\"columns\\"><div class=\\"column is-10 is-offset-1\\"><div class=\\"columns\\" id=\\"center-panel\\"><div class=\\"column\\"><input class=\\"has-text-centered\\" id=\\"search-text\\"></div><div class=\\"column is-narrow\\"><div class=\\"level\\" id=\\"center-panel\\"><div class=\\"level-item\\"><button id=\\"search-btn\\">Search</button><button id=\\"random-btn\\">Random</button></div></div></div></div></div></div><div class=\\"columns error hidden\\"><div class=\\"column\\"><div class=\\"level\\"><div class=\\"level-item\\"><div>An Error has occurred. Please check your internet connection and try again</div></div></div></div></div><div class=\\"columns spinner hidden\\"><img class=\\"column is-6 is-offset-3\\" style=\\"margin: 10vh auto 0 auto;\\"></div><div class=\\"columns is-mobile is-multiline hidden\\" id=\\"content\\"></div></div></section></div></body>"`;
